astcontext/blocks.go

```diff
@@ -0,0 +1,221 @@
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.SwitchStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Switch)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.FuncDecl:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Type.Func)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.FuncLit:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Type.Func)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		case *ast.TypeSwitchStmt:
+			bl := &Block{
+				BlockPos: ToPosition(p.fset.Position(x.Switch)),
+				node:     x,
+			}
+
+			if x.Body != nil {
+				bl.Lbrace = ToPosition(p.fset.Position(x.Body.Lbrace))
+				bl.Rbrace = ToPosition(p.fset.Position(x.Body.Rbrace))
+			}
+			blocks = append(blocks, bl)
+
+		}
+		return true
+	}
+
+	for _, file := range files {
+		// Inspect the AST and find all function declarements and literals
+		ast.Inspect(file, inspect)
+	}
+
+	return blocks
+}
+
+// BorderType represents Block's border type: left or right.
+type borderType string
+
+const (
+	leftBorderType  borderType = "left"
+	rightBorderType borderType = "right"
+)
+
+// Border is helper struct, that represents Block's border
+type border struct {
+	Block *Block
+	Pos   *Position
+	Type  borderType
+}
+
+// Borders represent a list of Border's
+type borders []*border
+
+// Borders returns Borders for Blocks
+func (b Blocks) borders() borders {
+	var borders []*border
+	for _, block := range b {
+		bdl := &border{
+			Block: block,
+			Pos:   block.Lbrace,
+			Type:  leftBorderType,
+		}
+		bdr := &border{
+			Block: block,
+			Pos:   block.Rbrace,
+			Type:  rightBorderType,
+		}
+		borders = append(borders, bdl, bdr)
+	}
+
+	return borders
+}
+
+// Sort implementation for Borders.
+func (b borders) Len() int      { return len(b) }
+func (b borders) Swap(i, j int) { b[i], b[j] = b[j], b[i] }
+func (b borders) Less(i, j int) bool {
+	return b[i].Pos.Offset < b[j].Pos.Offset
+}
+
+// FindBlock returns innermost Block within Borders for offset,
+// or error, if there is no such Block.
+func (b borders) findBlock(offset int) (*Block, error) {
+	sort.Sort(sort.Reverse(b))
+	rightTypes := 0
+
+	for _, border := range b {
+		// skip borders, that greater than offset
+		if border.Pos.Offset >= offset {
+			continue
+		}
+
+		if border.Type == leftBorderType {
+			if rightTypes == 0 {
+				return border.Block, nil
+			}
+			rightTypes--
+		} else {
+			rightTypes++
+		}
+	}
```

Comment:
Can't you just test it like this: https://github.com/fatih/motion/blob/master/astcontext/funcs.go#L271 ?